### PR TITLE
Configurable startup time

### PR DESCRIPTION
**Pull request by Isalia20**
## Description

Adds feature from  #3261  and from #2452

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

- [x] Test for long startup while having small response timeout. Test can be found in test/pytest/test_startup_timeout.py

================================================================================ test session starts =================================================================================
platform linux -- Python 3.10.14, pytest-7.3.1, pluggy-1.0.0
rootdir: /home/isalia/Desktop/torchserve
plugins: mock-3.14.0, timeout-2.3.1, cov-4.1.0, typeguard-4.3.0
collected 1 item                                                                                                                                                                     

test/pytest/test_startup_timeout.py .                                                                                                                                          [100%]

================================================================================= 1 passed in 15.02s =================================================================================

## Checklist:

- [x] Did you have fun?
- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: test/pytest/test_startup_timeout.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            if model_status == "READY":
+                break
+        time.sleep(1)
+
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+
+    assert response.status_code == 200, "Model startup failed"
+    assert model_status == "READY", f"Unexpected model status: {model_status}"
+    assert 10 <= elapsed_time < 20, f"Unexpected startup time: {elapsed_time} seconds"
```
Comment (mreso) [inline test/pytest/test_startup_timeout.py:116]: Currently, this test only tests the successful side of registering a model. Lets also test the situation where we run into the timeout and the worker gets killed. We can just configure a 5 seconds timeout and let the init sleep 10 as its now. We can parameterize the test for this. Then we can check in the stdout of TorchServe for the right message. See [here](https://github.com/pytorch/serve/blob/3f401806742a9430f85aabba7f2fa90a7be8c56d/test/pytest/test_continuous_batching.py#L92) how to get torchserve output in a test (The tests actually ignored it but the torchserve fixture is actually a Queue that you can .get() log messaged from). Let me know if you need help with this.

**Review by mreso — CHANGES_REQUESTED** [review]
Thanks for the contribution. Please see my comment regarding testing. Otherwise this look good to go after adjusting the test.

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java [hunk 2]
```diff
@@ -292,8 +297,8 @@ public void run() {
                 logger.debug("Shutting down the thread .. Scaling down.");
             } else {
                 logger.debug(
-                        "Backend worker monitoring thread interrupted or backend worker process died., responseTimeout:"
-                                + responseTimeout
+                        "Backend worker monitoring thread interrupted or backend worker process died., startupTimeout:"
```
[the remark this answers is not in the document]
Comment (Isalia20) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java:301]: makes sense, added the conditional

File: test/pytest/test_startup_timeout.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            model_status = response.json()[0]["workers"][0]["status"]
+            if model_status == "READY":
+                break
+        time.sleep(1)
+
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+
+    assert response.status_code == 200, "Model startup failed"
+    assert model_status == "READY", f"Unexpected model status: {model_status}"
```
[the remark this answers is not in the document]
Comment (Isalia20) [reply test/pytest/test_startup_timeout.py:139]: oops, missed that. Added the default value for it

File: test/pytest/test_startup_timeout.py (hunk 1 above)
Comment (Isalia20) [reply test/pytest/test_startup_timeout.py:140]: I've added the tests for it but not sure if it's the best way to do it. It reads the value from torchserve stdout and expects an error message to pop up and passes the test in that case.

Comment (mreso) [reply test/pytest/test_startup_timeout.py:140]: Great! That will do, thanks!

File: ts_scripts/spellcheck_conf/wordlist.txt [hunk 4]
```diff
@@ -218,6 +218,7 @@ preflight
 readthedocs
 req
 responseTimeout
+startupTimeout
```
Comment (mreso) [inline ts_scripts/spellcheck_conf/wordlist.txt:221]:
```suggestion
startupTimeout
timeframe
```

**Comment by Isalia20** [comment]
I see some checks are failing but I'm not sure what the issue really is, to fix it

**Comment by mreso** [comment]
> I see some checks are failing but I'm not sure what the issue really is, to fix it

Yeah, these regression tests can be a pain. Taking a look

**Comment by Isalia20** [comment]
One of the tests was failing due to startup_timeout not being cast to int in `test/pytest/test_gRPC_management_apis.py`. Added it to the list of params that should be casted to int

**Comment by Isalia20** [comment]
Seems like tests have passed(required ones). Docker ones are failing due to:
0.811 fatal: Remote branch configurable-startup-time not found in upstream origin
which could be because it can't clone from my branch 🤔

**Comment by mreso** [comment]
Hi @Isalia20 yes, that should be the docker build script which only supports pytorch/serve as remote. We can proceed with merging this PR. Thanks again for your contribution!

**Review by mreso — APPROVED** [review]
LGTM now

**Pull request merged**